src/toolbars/AudioSetupToolBar.cpp

```diff
@@ -0,0 +1,840 @@
+
+void AudioSetupToolBar::OnMenu(wxCommandEvent& event)
+{
+   int id = event.GetId();
+   bool audioSettingsChosen = false;
+
+   if ((id >= kHost) && (id < kInput)) {
+      ChangeHost(id);
+   }
+   else if ((id >= kInputChannels) && (id < kOutput)) {
+      if (auto item = mInputChannels->FindChildItem(id)) {
+         // Update cache with selected number of input channels
+         item->Check();
+         AudioIORecordChannels.Write(id - kInputChannels + 1);
+      }
+   }
+   else if ((id >= kInput) && (id < kInputChannels)) {
+      ChangeDevice(id, true);
+   }
+   else if ((id >= kOutput) && (id < kAudioSettings)) {
+      ChangeDevice(id, false);
+   }
+   else if (id == kAudioSettings) {
+      audioSettingsChosen = true;
+   }
+
+   auto gAudioIO = AudioIOBase::Get();
+   if (gAudioIO) {
+      // We cannot have gotten here if gAudioIO->IsAudioTokenActive(),
+      // per the setting of AudioIONotBusyFlag and AudioIOBusyFlag in
+      // AudacityProject::GetUpdateFlags().
+      // However, we can have an invalid audio token (so IsAudioTokenActive()
+      // is false), but be monitoring.
+      // If monitoring, have to stop the stream, so HandleDeviceChange() can work.
+      // We could disable the Preferences command while monitoring, i.e.,
+      // set AudioIONotBusyFlag/AudioIOBusyFlag according to monitoring, as well.
+      // Instead allow it because unlike recording, for example, monitoring
+      // is not clearly something that should prohibit changing device.
+      // TODO: We *could* be smarter in this method and call HandleDeviceChange()
+      // only when the device choices actually changed. True of lots of prefs!
+      // As is, we always stop monitoring before handling the device change.
+      if (gAudioIO->IsMonitoring())
+      {
+         gAudioIO->StopStream();
+         while (gAudioIO->IsBusy()) {
+            using namespace std::chrono;
+            std::this_thread::sleep_for(100ms);
+         }
+      }
+
+      if (audioSettingsChosen) {
+         if (auto project = ::GetActiveProject().lock()) {
+            PrefsPanel::Factories factories;
+            factories.push_back(PrefsPanel::PrefsNode(DevicePrefsFactory));
+
+            ViewDeviceSettingsDialog dialog(&GetProjectFrame(*project), *project, XO("Audio Settings:"), factories, 0);
+            dialog.SetSize(600, 420);
+            dialog.Center();
+
+            if (0 != dialog.ShowModal()) {
+               PrefsListener::Broadcast(DeviceToolbarPrefsID());
+            }
+         }
+      }
+      else {
+         gAudioIO->HandleDeviceChange();
+         PrefsListener::Broadcast(DeviceToolbarPrefsID());
+      }
+   }
+}
+
+static RegisteredToolbarFactory factory{ AudioSetupBarID,
+   []( AudacityProject &project ){
+      return ToolBar::Holder{ safenew AudioSetupToolBar{ project } };
+   }
+};
+
+namespace {
+AttachedToolBarMenuItem sAttachment{
+   /* i18n-hint: Clicking this menu item shows the toolbar
+      that manages devices */
```

Comment:
i18n (internationalization) hint comments are significant.  They get extracted into translation catalog template files and read by human experts in foreign languages.

Don't just copy-past this from DeviceToolBar.cpp.  Adapt it meaningfully.